source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2058 @@
+                    listener->accept_result_fn(listener, task_args->error_code, new_socket, task_args->user_data);
+                } else // The connection is not sent to user, clean it up. The nw_connection should be released in
+                       // socket clean up.
+                {
+                    aws_socket_clean_up(new_socket);
+                }
+            }
+            aws_mutex_unlock(&listener_nw_socket->synced_data.lock);
+        }
+    } else {
+        // If the task is not scheduled, release the connection.
+        nw_release(task_args->new_connection);
+    }
+
+    s_socket_release_internal_ref(listener_nw_socket);
+
+    aws_mem_release(task_args->allocator, task);
+    aws_mem_release(task_args->allocator, task_args);
+}
+
+static void s_schedule_on_listener_success(
+    struct nw_socket *nw_socket,
+    int error_code,
+    nw_connection_t new_connection,
+    void *user_data) {
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+    if (nw_socket->synced_data.base_socket && s_validate_event_loop(nw_socket->event_loop)) {
+        struct aws_task *task = aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct aws_task));
+
+        struct nw_listener_connection_args *args =
+            aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct nw_listener_connection_args));
+
+        args->nw_socket = nw_socket;
+        args->allocator = nw_socket->allocator;
+        args->error_code = error_code;
+        args->new_connection = new_connection;
+        args->user_data = user_data;
+
+        s_socket_acquire_internal_ref(nw_socket);
+        nw_retain(new_connection);
+
+        aws_task_init(task, s_process_listener_success_task, args, "listenerSuccessTask");
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, task);
+    }
+    aws_mutex_unlock(&nw_socket->synced_data.lock);
+}
+
+static void s_process_write_task(struct aws_task *task, void *args, enum aws_task_status status) {
+
+    struct nw_socket_written_args *task_args = args;
+    struct aws_allocator *allocator = task_args->allocator;
+    struct nw_socket *nw_socket = task_args->nw_socket;
+
+    if (status != AWS_TASK_STATUS_CANCELED) {
+        aws_mutex_lock(&nw_socket->synced_data.lock);
+        struct aws_socket *socket = nw_socket->synced_data.base_socket;
+        if (task_args->written_fn) {
+            task_args->written_fn(socket, task_args->error_code, task_args->bytes_written, task_args->user_data);
+        }
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+    }
+
+    s_socket_release_internal_ref(nw_socket);
+
+    aws_mem_release(allocator, task);
+    aws_mem_release(allocator, task_args);
+}
+
+static void s_schedule_write_fn(
+    struct nw_socket *nw_socket,
+    int error_code,
+    size_t bytes_written,
+    void *user_data,
+    aws_socket_on_write_completed_fn *written_fn) {
+    aws_mutex_lock(&nw_socket->synced_data.lock);
+    if (s_validate_event_loop(nw_socket->event_loop)) {
```

Comment:
Is it possible to have completed a write to a socket that does not have an event loop?  By the time we're reading/writing hasn't everything been set?